library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,33 @@
+[JLBP-4] Avoid dependencies known to cause runtime conflicts
+------------------------------
+
+- Avoid using dependencies that are not stable.
```

Comment:
"Stable" feels like not quite the right word for these 2 cases. Presumably App Engine isn't going to change its version of JCache, nor is JSR 305 going to remove classes.

Maybe these are just 2 additional top-level bullets. Actually, the App Engine one sounds like another example of the existing "Avoid using dependencies that duplicate classes with other dependencies" bullet (if I'm understanding right). So that would just leave a bullet about dependencies that share packages with other dependencies. (That's almost a generalization of "duplicate classes with other dependencies" to cover packages, too.)

Reply:
The mention of JCache is more about App Engine's mistake, it isn't about App Engine stability. Because of this mistake, and App Engine's promise not to break surface, it is stuck on the unstable JCache version. I'm guessing that both App Engine's version of JCache and the final version use the same artifact name; if so, this is not an example of multiple dependencies providing duplicate classes, it would be a single artifact that Maven can choose one of (but which results in conflicts for someone). Should more details be given on this point?

As for JSR 305 - you're probably right that jsr 305 probably won't be changed to break people, although the proposal is technically dormant and not final. But then again, the findbugs implementation is a stable version (3.x), so it might not make sense to worry about this aspect of it.

The other point about JSR 305 is the issue with the module system. @elharo or @saturnism , will this result in runtime conflicts?

Reply:
Oh, sorry, I'm misreading the text: "Non-final" doesn't mean "They made their own copy that removed the `final` keyword"; it just means "They added a dependency on an unstable version of JCache." Right?

This seems obvious in hindsight, but I think I was thrown by the difference in phrasing: The top-level bullet talks about "unstable" and "dependencies"; the sub-bullet talks about "non-final" and "classpath."

And as you note, I think I was also confused because I thought this was a recommendation to avoid App Engine, not a recommendation to avoid unstable JCache. (I mean, I guess it's sort of an indirect recommendation to avoid App Engine, to avoid these problems, but that's not the point :))

This also helps me understand why you refer to JSR305 here: As a non-accepted JSR, it could in theory change like the JCache API did.

I will think about this some more and try to figure out if there's something that could make it clearer or if that's not necessary and I'm just biased to read "non-final" as "not having the Java `final` keywords" because of my particular background :)

Reply:
Another possible part of the confusion: The early JCache API probably wouldn't have been "known to cause classpath conflicts" (per the title of this page) when App Engine added a dependency on it. The problem was likely that they didn't put enough weight on its _potential_, as an unstable API, to cause classpath conflicts later. Can we change the title to something that emphasizes that we should be anticipating particular problems (as the doc in fact says, contrary to my initial reading of this part :)), not just reacting after the fact?

Reply:
(It probably wasn't also clear to me that JCache was a JSR.)

Reply:
Anyway:

Possible page title: "Avoid dependencies that may cause runtime conflicts" (or, if you prefer, "Avoid dependencies that may conflict with other dependencies")

Possible first bullet and sub-bullets:

- Avoid using dependencies that are not stable.
  - For semantically versioned dependencies, this means avoiding both 0.x libraries and libraries that release many major versions. \[maybe a note about Guava's version numbering?\]
  - Avoid using unstable APIs \[link to JLBP-3?\] (except from other modules in your own library).
  - For JSRs, avoid using them until they are final.
    - The current exception is JSR 305 ...
    - But dependencies on other JSRs have caused real problems. For example, App Engine began to depend on JSR-107 (JCache) before it was finalized, and now the version it requires conflicts with the final version.

Reply:
(We might also just omit the App Engine example. I'd like to think that, of all the things we're saying here, our target audience understands why depending on unstable libraries is bad :) Or if they don't, I'm not sure that a one-sentence example is going to change that.)